dali/operators/random/rng_base_gpu.cuh

```diff
@@ -28,33 +28,82 @@ namespace dali {
 
 namespace {
 
+template <bool value>
+using bool_const = std::integral_constant<bool, value>;
+
 template <typename T, typename Dist>
-__device__ __inline__ void Generate(BlockDesc<true> desc,
+__device__ __inline__ void Generate(const SampleDesc &sample,
+                                    const BlockDesc &block,
                                     Dist& dist,
-                                    curandState* __restrict__ rng) {
-  auto out = static_cast<T*>(desc.output);
-  auto in = static_cast<const T*>(desc.input);
-  auto idx_end = desc.offset + desc.size;
-  for (auto idx = desc.offset + threadIdx.x; idx < idx_end; idx += blockDim.x) {
+                                    curandState* __restrict__ rng,
+                                    bool_const<true>,     // is_noise_gen
+                                    bool_const<true>) {   // is_per_channel
+  auto out = static_cast<T*>(sample.output);
+  auto in = static_cast<const T*>(sample.input);
+  auto idx_end = block.p_offset + block.p_count;
+  for (auto idx = block.p_offset + threadIdx.x; idx < idx_end; idx += blockDim.x) {
     auto n = dist.Generate(in[idx], rng);
     dist.Apply(out[idx], in[idx], n);
   }
 }
 
 template <typename T, typename Dist>
-__device__ __inline__ void Generate(BlockDesc<false> desc,
+__device__ __inline__ void Generate(const SampleDesc &sample,
+                                    const BlockDesc &block,
+                                    Dist& dist,
+                                    curandState* __restrict__ rng,
+                                    bool_const<true>,     // is_noise_gen
+                                    bool_const<false>) {  // is_per_channel
+  auto out = static_cast<T*>(sample.output);
+  auto in = static_cast<const T*>(sample.input);
+  auto idx_end = block.p_offset + block.p_count;
+  for (auto idx = block.p_offset + threadIdx.x; idx < idx_end; idx += blockDim.x) {
+    int64_t pos = idx * sample.p_stride;
+    // Implementations that generate noise once for all channels should not depend on the input
+    // to generate the number.
+    auto n = dist.Generate({}, rng);
+    for (int c = 0; c < sample.c_count; c++, pos += sample.c_stride) {
+      dist.Apply(out[pos], in[pos], n);
+    }
+  }
+}
+
+template <typename T, typename Dist>
+__device__ __inline__ void Generate(const SampleDesc &sample,
+                                    const BlockDesc &block,
+                                    Dist& dist,
+                                    curandState* __restrict__ rng,
+                                    bool_const<false>,     // is_noise_gen
+                                    bool_const<true>) {    // is_per_channel
+  auto out = static_cast<T*>(sample.output);
+  auto idx_end = block.p_offset + block.p_count;
+  for (auto idx = block.p_offset + threadIdx.x; idx < idx_end; idx += blockDim.x) {
+    auto n = dist.Generate(rng);
+    out[idx] = ConvertSat<T>(n);
+  }
+}
+
+template <typename T, typename Dist>
```

Comment:
4. is_noise_gen=False, is_per_channel=False.
RNG (generates a random number, doesn't depend on any input) that produces the same number for all channels in a pixel